File: data_structures/graph/dfs.py
"""
Given a connected undirected graph represented by an adjacency list adj, which is a vector of vectors where each adj[i] represents the list of vertices connected to vertex i. Perform a Depth First Traversal (DFS) starting from vertex 0, visiting vertices from left to right as per the adjacency list, and return a list containing the DFS traversal of the graph.

Note: Do traverse in the same order as they are in the adjacency list.

Examples:

Input: adj = [[2,3,1], [0], [0,4], [0], [2]]

Output: [0, 2, 4, 3, 1]
Explanation: Starting from 0, the DFS traversal proceeds as follows: 
Visit 0 → Output: 0 
Visit 2 (the first neighbor of 0) → Output: 0, 2 
Visit 4 (the first neighbor of 2) → Output: 0, 2, 4 
Backtrack to 2, then backtrack to 0, and visit 3 → Output: 0, 2, 4, 3 
Finally, backtrack to 0 and visit 1 → Final Output: 0, 2, 4, 3, 1
"""

class Solution:
    
    #Function to return a list containing the DFS traversal of the graph.
    def dfsOfGraph(self, adj):
        # code here
        visited = [False] * (len(adj) + 1)
        dfs_list = []
        start_node = 0
        self.dfs(start_node, adj, dfs_list, visited)
        return dfs_list
    
    def dfs(self, node, adj, dfs_list, visited):
        visited[node] = True
        dfs_list.append(node)
        
        for neighbor in adj[node]:
            if not visited[neighbor]:
                self.dfs(neighbor, adj, dfs_list, visited)
    

sample = [[2,3,1], [0], [0,4], [0], [2]]
print("solution:", Solution().dfsOfGraph(sample))

File: data_structures/graph/dfs.ts
class Solution{
    dfsOfGraph(adj: number[][]): number[]{
        const visited: boolean[] = [];
        for (let i=0; i<adj.length; i++) {
            visited[i] = false;
        }
        const dfsList: number[] = [];
        const start_node: number = 0;

        this.dfs(start_node, adj, dfsList, visited);

        return dfsList;
    }

    private dfs(node: number, adj: number[][], dfsList: number[], visited: boolean[]){
        visited[node] = true;
        dfsList.push(node);

        for (const neighbor of adj[node]){
            if (!visited[neighbor]) {
                this.dfs(neighbor, adj, dfsList, visited);
            }
        }


    }
}

// Example usage:
const graph: number[][] = [
    [1, 2],    // Node 0 is connected to 1 and 2
    [0, 3, 4], // Node 1 is connected to 0, 3, and 4
    [0, 5, 6], // Node 2 is connected to 0, 5, and 6
    [1],       // Node 3 is connected to 1
    [1],       // Node 4 is connected to 1
    [2],       // Node 5 is connected to 2
    [2]        // Node 6 is connected to 2
];
const solution = new Solution();
console.log("ans is:",solution.dfsOfGraph(graph).toString());